File: modules/CalicoParser/src/CalicoParserModule.ts
import { NativeModule, requireNativeModule } from 'expo';
import { Metadata } from './CalicoParser.types';

declare class CalicoParserModule extends NativeModule {
    parseChapter(filePath: string, chapterPath: string): Promise<string>;
    importMetadata(filePath: string): Promise<Metadata>;
    chunkEpub(filePath: string): Promise<Record<string, string[]>>;
    parseChunk(filePath: string, chapterPaths: Array<string>): Promise<Record<string, string>>;
    clean(): void;
}

const nativeModule = requireNativeModule<CalicoParserModule>('CalicoParser');

export default {
    async parseChapter(filePath: string, chapterPath: string): Promise<string> {
        const result = await nativeModule.parseChapter(filePath, chapterPath);
        return result;
    },

    async parseChunk(filePath: string, chapterPaths: Array<string>): Promise<Record<string, string>> {
        const result = await nativeModule.parseChunk(filePath, chapterPaths);
        return result;
    },

    async importMetadata(filePath: string): Promise<Metadata> {
        const result = await nativeModule.importMetadata(filePath);
        return result;
    },

    async chunkEpub(filePath: string): Promise<Record<string, string[]>> {
        const result = await nativeModule.chunkEpub(filePath);
        return result;
    },

    async clean(): Promise<void> {
        nativeModule.clean();
    },
};

File: modules/CalicoParser/android/src/main/java/expo/module/calicoparser/CalicoParserModule.kt
package expo.module.calicoparser

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import android.net.Uri

import expo.module.calicoparser.epub.EpubParser

class CalicoParserModule : Module() {
    private val epubParser = EpubParser()

    override fun definition() = ModuleDefinition {
        Name("CalicoParser")

        AsyncFunction("clean") { uriString: String ->
            epubParser.cleanCache()
        }

        AsyncFunction("parseChapter") { uriString: String, chapterPath: String ->
            val context = appContext.reactContext ?: throw Exception("Context not available")
            val uri = Uri.parse(uriString)
            return@AsyncFunction epubParser.parseChapter(context, uri, chapterPath)
        }

        AsyncFunction("parseChunk") { uriString: String, chapterPaths: List<String> ->
            val context = appContext.reactContext ?: throw Exception("Context not available")
            val uri = Uri.parse(uriString)
            return@AsyncFunction epubParser.parseChunk(context, uri, chapterPaths)
        }

        AsyncFunction("importMetadata") { uriString: String ->
            val context = appContext.reactContext ?: throw Exception("Context not available")
            val uri = Uri.parse(uriString)
            epubParser.importMetadata(context, uri)
        }

        AsyncFunction("chunkEpub") { uriString: String ->
            val context = appContext.reactContext ?: throw Exception("Context not available")
            val uri = Uri.parse(uriString)
            epubParser.chunkEpub(context, uri)
        }
    }
}
